### PR TITLE
Parse sexpr bir

### LINKED ISSUE #1031: Implement bir textual format
Design a textual format to represent bir. Implement a serializer and a deserializer for it.

### PR DESCRIPTION
**Pull request by manuranga**
Fixes #1031

### CODE REVIEW

File: compiler/tests/testSuite.bal [hunk 1]
```diff
@@ -283,3 +288,30 @@ function testListAllMemberTypesProjection(t:Context tc, t:SemType t) {
         }
     }
 }
+
```
Comment (manuranga) [inline compiler/tests/testSuite.bal:291]: Add a comment

File: compiler/modules/bir.sexpr/parse.bal [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+|};
+
+type FuncParseContext record {|
+    *ParseContext;
+    map<bir:Register> regs = {};
+    map<bir:Label> blockLabels;
+|};
+
+const DEFAULT_ROOT_MODULE_NAME = "root";
+final bir:ModuleId DEFAULT_ROOT_MODULE_ID = { org: "", names: [DEFAULT_ROOT_MODULE_NAME] };
```
Comment (manuranga) [inline compiler/modules/bir.sexpr/parse.bal]: remove

**Comment by manuranga** [comment]
I am considering how to provide `File` abstraction on top of bir, since `lineContent` and `qNameRange` APIs are difficult to implement on top of sexpr (without more of metadata). `lineContent` can be refactored away since it's only need in `front`. `qNameRange` is more problematic since it's need by the verifier.
Option 1. Attach `qNameRange` to meta data.
```
(meta (parts "mul-div.bal")
      (qname-ranges ((line 1 3) (line 1 10))
                    ((line 3 2) (line 3 8))))
```
Option 2. Remove `qNameRange` and add something like `nameEndPos` to AST.

**Comment by manuranga** [comment]
>  `lineContent` can be refactored away

This turned out to be too optimistic, Don't have access to `SourceFile` even from front since we access it via `CompileError`. Can do a `is` check, but ugly.

**Comment by manuranga** [comment]
Todo (discussed in Mar 20 meeting):

- [x] change line to loc
- [x] remove meta
- [x] change the file to key based.
- [x] change defn to defns
- [x] get rid of $ in narrow regs
- [ ] change type-br to type-cond-br

**Review by jclark — APPROVED** [review]

**Pull request merged**